test/jdk/java/lang/Thread/crac/JoinOrSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,194 @@
+                        } else {
+                            mainThread.join(T_MS);
+                        }
+                        break;
+
+                    case sleep:
+                        if (nsAccuracy) {
+                            Thread.sleep(T_MS, T_NS);
+                        } else {
+                            Thread.sleep(T_MS);
+                        }
+                        break;
+
+                    default:
+                        throw new IllegalArgumentException("unknown test type");
+                }
+
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+
+            tDone = System.nanoTime();
+        };
+
+        Thread t = new Thread(r);
+        t.start();
+
+        // this additional synchronization is probably redundant;
+        // adding it to ensure we get the expected TIMED_WAITING state
+        // from "our" join or sleep
+        checkpointLatch.await();
+
+        // it is expected that EPS_NS is enough to complete the join/sleep
+        // on restore => expecting that 5 * EPS_NS is enough to enter them
+        long dt = 5 * EPS_NS;
+        Thread.sleep(dt / 1_000_000);
+
+        if (t.getState() != Thread.State.TIMED_WAITING) {
+            throw new RuntimeException("was not able to enter " + op
+                + " in " + dt + " ns");
+        }
+
+        long tBeforeCheckpoint = System.nanoTime();
+
+        jdk.crac.Core.checkpointRestore();
+
+        long tAfterRestore = System.nanoTime();
+
+        t.join();
+
+        long pause = tAfterRestore - tBeforeCheckpoint;
+        if (pause < 1_000_000 * CRPAUSE_MS - EPS_NS) {
+            throw new RuntimeException(
+                "the CR pause was less than " + CRPAUSE_MS + " ms");
+        }
+
+        if (tDone < tBeforeCheckpoint + EPS_NS) {
+            throw new RuntimeException(
+                op + " has finished before the checkpoint");
+        }
+
+        long eps = Math.abs(tAfterRestore - tDone);
+
+        if (eps > EPS_NS) {
+            throw new RuntimeException(
+                "the " + op + "ing thread has finished in " + eps + " ns "
+                + "after the restore (expected: " + EPS_NS + " ns)");
+        }
+    }
+
+
+    public static void main(String args[]) throws Exception {
+
+        if (args.length > 0) {
+
+            if (args.length < 2) {
+                throw new RuntimeException(
+                    "please specify test parameters (testType, nsAccuracy)");
+            }
+
+            new JoinOrSleepOnCRPauseTest(
+                    TestType.valueOf(args[0]),
+                    Boolean.parseBoolean(args[1])).runTest();
+
+        } else {
+
+            ProcessBuilder pb;
+            OutputAnalyzer out;
+
+            String type[] = {"join", "sleep"};
+            String nsAccuracy[] = {"false", "true"};
+
+            for (String t: type) {
+                for (String acc: nsAccuracy) {
```

Comment:
Please provide these as the argument to the test
```
* @run main/othervm JoinOrSleepOnCRPauseTest sleep_ms
* @run main/othervm JoinOrSleepOnCRPauseTest sleep_ns
* @run main/othervm JoinOrSleepOnCRPauseTest wait_ns
```

Reply:
done